### DISCUSSION TITLE
Error Using Client only Package in SSR

### DISCUSSION THREAD

**Discussion by Zibbp** [nuxt/nuxt#16541] [Questions] [answered]
Hi, 

I'm attempting to use [Plyr](https://github.com/sampotts/plyr) in a Nuxt3 SSR project but all attempts lead to a `document is not defined` error. I've tried wrapping the player component in `<ClientOnly>` but that did not work. 

Is there a way to force the import and initialization of Plyr on the client while having the Nuxt3 application be SSR?

Sandbox link showing the issue: [https://codesandbox.io/s/wonderful-austin-bl4w89](https://codesandbox.io/s/wonderful-austin-bl4w89)

**Comment by Megarsh** [comment, marked as answer]
As far as I can tell the ClientOnly components are still processed on the server and that's why you're getting the undefined document error.

You can import it in onMounted, and with that I think you could also get rid of the ClientOnly component

```
onMounted(async () => {
    const Plyr = await import('plyr');

    const newPlayer = new Plyr.default('#player');
})
```

**Reply by Zibbp** [reply]
I added the dynamic import and moved the call to an onMounted hook but it doesn't appear to create a new Plyr instance on the video element. It isn't throwing any errors either.
https://codesandbox.io/s/agitated-yonath-yw7ug5?file=/components/Player.vue

**Reply by Megarsh** [reply]
Sorry forgot that it returns a promise, edited the answer it should work now

**Reply by Zibbp** [reply]
Works with the updated code, thanks!

**Reply by bennymi** [reply]
@Zibbp I'm currently also trying to use Plyr with Nuxt3 and I was able to embed a YouTube video with the Plyr default style from this post, but I'm still unable to control or customize the video from the Vue file. Were you able to figure out how to do it?

**Reply by Zibbp** [reply]
@bennymi Here is my working layout. I have an empty video element with an ID in the template as the video url is retrieved via an API.
```
<video id="player" playsinline controls crossorigin=""></video>
```

In the `<script setup>` I have two objects, playerOptions and playerSources which will be used when creating the Plyr instance.

```
const playerOptions = {
  controls: [
    "rewind",
    "play-large",
    "play",
    "fast-forward",
    "progress",
    "current-time",
    "mute",
    "volume",
    "captions",
    "settings",
    "download",
    "pip",
    "airplay",
    "fullscreen",
  ],
  settings: ["captions", "quality", "speed", "loop"],
  autoplay: false,
  keyboard: { focused: true, global: false },
  speed: { selected: 1, options: [0.5, 0.75, 1, 1.25, 1.5, 1.75, 2] },
  markers: {},
};
const playerSources = {
  type: "video",
  title: "Eos",
  sources: [],
  poster: "",
  tracks: [],
};
```
Below that, I have the `onMounted` function from the above comment. Inside this function resides the Plyr functions.

```
onMounted(async () => {
  // Import Plyr
  const Plyr = await import("plyr");
  
  // Create player instance with the options  from above
  const newPlayer = new Plyr.default("#player", playerOptions);
  
  // Push the video source retrieved via an API to the playerSources object 
  playerSources.sources.push({
    src: `${config.cdnURL}/${encodeURIComponent(props.video.video_path)}`,
    type: "video/webm",
  });
 
  // Set player sources using the playerSources object
  newPlayer.source = playerSources;
});
```
Hope this helps.

**Reply by valentinhttr** [reply]
Hi, I am trying to do the same thing with pdf_viewer from `pdfjs-dist`.

It was working on Nuxt 2, but not on Nuxt Bridge :
```js
onMounted(async () => {
    const pdfViewer = await import('pdfjs-dist/legacy/web/pdf_viewer.js')
})
```
When building to production `npm run build` and trying it, i get `navigator is not defined`.

Do you know why the first one no longer works please?
```js
onMounted(async () => {
   if(process.browser) {
       const pdfViewer = await import('pdfjs-dist/legacy/web/pdf_viewer.js')
   }
})
```

**Comment by sewalsh** [comment]
Not your solution but here's a sandbox with [VidStack](https://github.com/vidstack/vidstack) (which Plyr is being merged into).

https://codesandbox.io/s/awesome-sea-xk5xrf?file=/pages/index.vue

**Reply by Zibbp** [reply]
I saw that a few Players were getting merged into Vidstack, it sounds promising! I tried to view your example but I received some errors and it's displaying the default browser video element.

**Reply by sewalsh** [reply]
The default out of the box skin is very similar to a HTML5 video but the overflow menu should reveal it's using VidStack.

<img width="557" alt="image" src="https://user-images.githubusercontent.com/7760319/167221364-39377016-c2a8-42d5-a224-6985632605b0.png">

**Comment by danielroe** [comment]
Nuxt supports client only components that don't get included in the server build. Try creating a `VideoPlayer.client.vue` and using plyr only within that component.

**Reply by Zibbp** [reply]
I tried using a `*.client.vue` component but received an error using a normal `import package from "package"` and calling it.

```
Unhandled error during execution of render function 
  at <ClientOnlyWrapper> 
  at <Index onVnodeUnmounted=fn<onVnodeUnmounted> ref=Ref< undefined > key="/" > 
  at <BaseTransition mode="out-in" appear=false persisted=false  ... > 
  at <Transition name="page" mode="out-in" > 
  at <RouterView > 
  at <NuxtPage> 
  at <App key=1 > 
  at <NuxtRoot>
```

@Megarsh 's answer is working perfectly.

**Reply by treighmawaka** [reply]
> Nuxt supports client only components that don't get included in the server build. Try creating a `VideoPlayer.client.vue` and using plyr only within that component.

Great tip! This works well for standard components. It should be documented.
